asm/exceptasm.s

```diff
@@ -69,646 +81,897 @@ glabel __osIntTable
-/* 004A18 80003E18 11400003 */  beqz  $t2, firstnmi
-/* 004A1C 80003E1C 2401EFFF */   li    $at, -4097
-/* 004A20 80003E20 10000023 */  b     redispatch
-/* 004A24 80003E24 02018024 */   and   $s0, $s0, $at
+    lw      $k1, THREAD_SR($k0)
+    # Clear interrupt
+    li      $at, ~CAUSE_IP5
+    lui     $t1, %hi(__osShutdown)
+    and     $k1, $k1, $at
+    sw      $k1, THREAD_SR($k0)
+    addiu   $t1, %lo(__osShutdown)
+    # Test __osShutdown for first nmi
+    lw      $t2, ($t1)
+    beqz    $t2, firstnmi
+     li     $at, ~CAUSE_IP5
+    # Mask out interrupt and redispatch immediately
+    b       redispatch
+     and    $s0, $s0, $at
+
 firstnmi:
-/* 004A28 80003E28 240A0001 */  li    $t2, 1
-/* 004A2C 80003E2C AD2A0000 */  sw    $t2, ($t1)
-/* 004A30 80003E30 0C000FCB */  jal   send_mesg
-/* 004A34 80003E34 24040070 */   li    $a0, 112
-/* 004A38 80003E38 3C0A8001 */  lui   $t2, %hi(__osRunQueue) # $t2, 0x8001
-/* 004A3C 80003E3C 8D4AAD48 */  lw    $t2, %lo(__osRunQueue)($t2)
-/* 004A40 80003E40 2401EFFF */  li    $at, -4097
-/* 004A44 80003E44 02018024 */  and   $s0, $s0, $at
-/* 004A48 80003E48 8D5B0118 */  lw    $k1, 0x118($t2)
-/* 004A4C 80003E4C 0361D824 */  and   $k1, $k1, $at
-/* 004A50 80003E50 10000017 */  b     redispatch
-/* 004A54 80003E54 AD5B0118 */   sw    $k1, 0x118($t2)
+    # Set __osShutdown
+    li      $t2, 1
+    sw      $t2, ($t1)
+    # Post a PreNMI event message
+    jal     send_mesg
+     li     $a0, OS_EVENT_PRENMI*8
+    # Clear interrupt
+    lui     $t2, %hi(__osRunQueue)
+    lw      $t2, %lo(__osRunQueue)($t2)
+    li      $at, ~CAUSE_IP5
+    and     $s0, $s0, $at
+    lw      $k1, THREAD_SR($t2)
+    and     $k1, $k1, $at
+    # Redispatch immediately
+    b       redispatch
+     sw     $k1, THREAD_SR($t2)
+
 sw2:
-/* 004A58 80003E58 2401FDFF */  li    $at, -513
-/* 004A5C 80003E5C 01014024 */  and   $t0, $t0, $at
-/* 004A60 80003E60 40886800 */  mtc0  $t0, $13
-/* 004A64 80003E64 0C000FCB */  jal   send_mesg
-/* 004A68 80003E68 24040008 */   li    $a0, 8
-/* 004A6C 80003E6C 2401FDFF */  li    $at, -513
-/* 004A70 80003E70 1000FF5B */  b     next_interrupt
-/* 004A74 80003E74 02018024 */   and   $s0, $s0, $at
+    li      $at, ~CAUSE_SW2
+    # Clear interrupt
+    and     $t0, $t0, $at
+    mtc0    $t0, C0_CAUSE
+    # Post sw2 event message
+    jal     send_mesg
+     li     $a0, OS_EVENT_SW2*8
+    li      $at, ~CAUSE_SW2
+    # Mask out interrupt and continue
+    b       next_interrupt
+     and    $s0, $s0, $at
+
 sw1:
-/* 004A78 80003E78 2401FEFF */  li    $at, -257
-/* 004A7C 80003E7C 01014024 */  and   $t0, $t0, $at
-/* 004A80 80003E80 40886800 */  mtc0  $t0, $13
-/* 004A84 80003E84 0C000FCB */  jal   send_mesg
-/* 004A88 80003E88 24040000 */   li    $a0, 0
-/* 004A8C 80003E8C 2401FEFF */  li    $at, -257
-/* 004A90 80003E90 1000FF53 */  b     next_interrupt
-/* 004A94 80003E94 02018024 */   and   $s0, $s0, $at
+    li      $at, ~CAUSE_SW1
+    # Clear interrupt
```

Comment:
```suggestion
    # Clear interrupt
    li      $at, ~CAUSE_SW1
```
(same for sw2)


Also, is it guaranteed that t0 still hold Cause at this point? it's set to that in `handle_interrupt` but `rcp` also uses t0, and if both bits are set at the same time rcp runs before sw1 and sw2

So can two bits of the Cause interrupt pending mask be set at the same time?
The answer has to be yes since exceptasm accounts for that scenario (by branching back to `next_interrupt`)...

Reply:
Yeah this is definitely a bug.. however the chances of both interrupts happening in the timeframe needed to trigger it is small I suppose, and also the software interrupts are (almost?) never asserted anyway in any code I've ever looked at